File: singleton/Logger.py
from typing import Optional
from datetime import datetime


class Logger:

    _FORMAT = "%s - %s"
    _instance: Optional["Logger"] = None

    def __new__(cls):
        if Logger._instance is None:
            Logger._instance = object.__new__(cls)
        return Logger._instance

    def log(self, msg: str) -> None:
        print(Logger._FORMAT % (datetime.now(), msg))


if __name__ == "__main__":
    logger = Logger()
    logger.log("Logger initialized")
    assert logger == Logger()

File: singleton/Logger.java
import static java.lang.String.format;
import java.time.LocalDateTime;

class Logger {

    private static final String FMT = "%s - %s";
    private static Logger instance = null;
    
    private Logger() { }

    public static Logger getLogger() {
        if (instance == null) {
            instance = new Logger();
        }
        return instance;
    }

    public void log(String msg) {
        System.out.println(format(FMT, LocalDateTime.now(), msg));
    }
    
    public static void main(String args[]) {
        Logger logger = Logger.getLogger();
        logger.log("Logger initialized");
        assert(logger.equals(Logger.getLogger()));
    }

}